app/ide-desktop/lib/content/src/index.ts

```diff
@@ -119,104 +117,134 @@ function displayDeprecatedVersionDialog() {
-                    /** FIXME [PB]: https://github.com/enso-org/cloud-v2/issues/366
-                     * React hooks rerender themselves multiple times. It is resulting in multiple
-                     * Enso main scene being initialized. As a temporary workaround we check whether
-                     * appInstance was already ran. Target solution should move running appInstance
-                     * where it will be called only once. */
-                    let appInstanceRan = false
-                    const onAuthenticated = () => {
-                        if (
-                            !contentConfig.OPTIONS.groups.featurePreview.options.newDashboard.value
-                        ) {
-                            hideAuth()
-                            if (!appInstanceRan) {
-                                appInstanceRan = true
-                                void appInstance.run()
-                            }
-                        }
-                    }
-                    authentication.run({
-                        logger,
-                        platform,
-                        projectManager: projectManager.ProjectManager.default(),
-                        showDashboard:
-                            contentConfig.OPTIONS.groups.featurePreview.options.newDashboard.value,
-                        onAuthenticated,
-                    })
-                } else {
-                    void appInstance.run()
-                }
-                const email = contentConfig.OPTIONS.groups.authentication.options.email.value
-                // The default value is `""`, so a truthiness check is most appropriate here.
-                if (email) {
-                    logger.log(`User identified as '${email}'.`)
-                }
+            void appInstance.run()
+            dropPreviousInstance = () => {
+                // The `any` is unavoidable as `App.wasm` is typed as `any`.
+                // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
+                appInstance.wasm?.drop?.()
             }
-        } else {
-            console.error('Failed to initialize the application.')
         }
     }
 }
 
-const API = new Main()
+function main() {
+    // Load command line arguments into `configOptions.OPTIONS`.
+    new app.App({
+        configOptions: contentConfig.OPTIONS,
+    })
+
+    const isUsingAuthentication = contentConfig.OPTIONS.options.authentication.value
+    const isUsingNewDashboard =
+        contentConfig.OPTIONS.groups.featurePreview.options.newDashboard.value
+    const isOpeningIde =
+        contentConfig.OPTIONS.groups.startup.options.entry.value ===
+        contentConfig.OPTIONS.groups.startup.options.entry.default
+
+    if ((isUsingAuthentication || isUsingNewDashboard) && isOpeningIde) {
+        window.tryStopApp = tryStopApp
+        window.runApp = runApp
```

Comment:
Why are we polluting the window scope? Does Rust need to have access to it?

Reply:
not rust, but the dashboard does need access - the alternative to polluting the window scope is to basically pass `app` all the way through React (basically what `ide?: app.App` was for), but i think you had issues with doing that as well? so not sure what would be a satisfactory solution, unfortunately

although it's definitely not a good reason to do so, the previous implementation was already using a global variable.

Reply:
not rust, but the dashboard does need access - the alternative to polluting the window scope is to basically pass `app` all the way through React (basically what `ide?: app.App` was for), but i think there were architectural issues with doing that as well? so not sure if i have a satisfactory solution, unfortunately